### PR TITLE
Feat/paginated select

### PR DESCRIPTION
**Pull request by AlecAivazis**
This PR adds a few major improvements to the package in order to address a few issues that were recently reported as well as some things I ran into while writing those fixes. The important ones are:

1. Select and MultiSelect cycle through the top and bottom when the user presses the appropriate key [#71]
1. Select and MultiSelect are paginated [#70, #61] 
    1. default page size is 7
    1. both prompts have a `PageSize` field which allows for independent page sizes for prompts
    1. page size can be changed by setting `survey.PageSize` to another value
1. Improved the readline behavior to allow for the cursor to be moved with the arrows keys and insert text anywhere in the line

### CODE REVIEW

**Comment by coryb** [comment]
Hey @AlecAivazis, I will try to dig into this PR, there is a lot of changes to review here.  Not sure why the test failed also.

**Comment by AlecAivazis** [comment]
Thanks @coryb. Sorry about the high number of changes - I wanted to have it be 3 separate PRs but had very limited access over vacation and one thing led to another... I'm planning on investigating the failing test in the next few days

File: multiselect.go [hunk 1]
```diff
@@ -28,9 +29,10 @@ type MultiSelectTemplateData struct {
 	MultiSelect
 	Answer        string
 	ShowAnswer    bool
-	Checked       map[int]bool
+	Checked       map[string]bool
 	SelectedIndex int
 	ShowHelp      bool
+	Choices       []string
```
Comment (coryb) [inline multiselect.go]: Wondering if we should change the name of this, maybe `Page`, `PageOptions` or something like that?  `Choices` and `Options` are indistinguishable to me.

File: survey_test.go [hunk 2]
```diff
@@ -42,3 +42,97 @@ func TestAsk_returnsErrorIfTargetIsNil(t *testing.T) {
...
+	choices := []string{"choice1", "choice2", "choice3"}
+	// a page bigger than the total number
+	pageSize := 4
+	// the current selection
+	sel := 2
+
+	// compute the page info
+	choices, idx := paginate(pageSize, choices, sel)
+
+	if choices[0] != "choice1" && choices[1] != "choice2" && choices[2] != "choice3" {
```
Comment (coryb) [inline survey_test.go]: using testify like with other tests, I think this all simplifies to:

```
page, idx := paginate(pageSize, choices, sel)
assert.Equal(t, choices, page)
assert.Equal(t, 2, idx)
```

File: survey_test.go [hunk 3]
```diff
@@ -42,3 +42,97 @@ func TestAsk_returnsErrorIfTargetIsNil(t *testing.T) {
...
+	pageSize := 4
+
+	// test the second item
+	sel := 2
+
+	// compute the page info
+	choices, idx := paginate(pageSize, choices, sel)
+
+	// we should see the first three options
+	if choices[0] != "choice1" && choices[1] != "choice2" && choices[2] != "choice3" && choices[3] != "choice3" {
```
Comment (coryb) [inline survey_test.go]: again here simplify with testity, also fix logical error (need `||` vs `&&`)
```
page, idx := paginate(pageSize, choices, sel)
assert.Equal(t, choices[0:4], page)
assert.Equal(t, 2, idx)
```

File: survey_test.go [hunk 4]
```diff
@@ -42,3 +42,97 @@ func TestAsk_returnsErrorIfTargetIsNil(t *testing.T) {
...
+		t.Error("Did not recieve the correct index")
+	}
+
+}
+
+func TestPagination_firstHalf(t *testing.T) {
+	// the choices for the test
+	choices := []string{"choice1", "choice2", "choice3", "choice4", "choice5", "choice6"}
+
+	// section the choices into groups of 3
```
Comment (coryb) [inline survey_test.go]: comment does not match value, not sure which you intended

File: tests/select.go [hunk 5]
```diff
@@ -39,6 +39,18 @@ var goodTable = []TestUtil.TestTableEntry{
 			Options: []string{"red", "blue"},
 		}, &answer,
 	},
+	{
```
Comment (coryb) [inline tests/select.go]: you will need to regenerate the autoplay scripts since you changed the test

File: tests/selectThenInput.go [hunk 6]
```diff
@@ -16,13 +16,6 @@ var simpleQs = []*survey.Question{
 		},
 		Validate: survey.Required,
 	},
-	{
```
Comment (coryb) [inline tests/selectThenInput.go]: will need to update autoplay tests since you removed tests

File: select.go [hunk 7]
```diff
@@ -25,6 +26,7 @@ type Select struct {
 // the data available to the templates when processing
 type SelectTemplateData struct {
 	Select
+	Choices       []string
```
Comment (coryb) [inline select.go]: again here, I think we need to pick a different name, `Page` etc.

**Comment by coryb** [comment]
Hey, I am going to be out of town from Saturday thru Thursday with no internet access, so I am just going to approve this PR now.  Hopefully you can still get updates for the autoplay scripts.  The travis autoplay error is still confusing to me, it works on my osx laptop.  If they are still failing after you update for your changes then I will take a look, might be linux specific.

**Comment by AlecAivazis** [comment]
Awesome, thanks for looking at this @coryb  - I updated the autoplay tests and there seems to be an error in travis with a double carriage return on linux 😞

**Comment by coryb** [comment]
Okay, I confirmed the tests are working on OSX.  I think the issue is a timing issue on linux for some reason.  I can reproduce via a docker container, not sure why it so random.  
Anyway, to get travis working, it is probably best for now to just comment out the `run autoplay-tests` in the `.travis.yml` and we can try to get it more stable in the future.

**Comment by AlecAivazis** [comment]
yea I think you're right, i'm going to try to re-run the tests and if that doesn't work i'll comment out the line in the travis config - thanks for verifying that the autoplay tests pass for you!

**Pull request merged**